crates/gitbutler-repo/src/change_reference.rs

```diff
@@ -0,0 +1,213 @@
+    {
+        return Err(anyhow!(
+            "A reference {} already exists",
+            branch_reference.name
+        ));
+    }
+    // Ensure the change is not already referenced
+    if all_references.iter().any(|r| r.change_id == change_id) {
+        return Err(anyhow!(
+            "A reference for change {} already exists",
+            change_id
+        ));
+    }
+    validate_commit(&vbranch, commit.id(), ctx, &handle)?;
+    vbranch.references.push(branch_reference.clone());
+    handle.set_branch(vbranch)?;
+    Ok(branch_reference)
+}
+
+/// Updates an existing branch reference to point to a different commit.
+/// Only the commit and change_id can be updated.
+/// The reference is identified by the branch id and the reference name.
+/// This function will return an error if:
+/// - this reference does not exist
+/// - the reference exists, but the commit id is not in the branch
+/// - the reference exists, but the commit id is already associated with another reference
+///
+/// If the commit ID is the same as the current commit ID, the function is a no-op.
+pub fn update_change_reference(
+    ctx: &CommandContext,
+    branch_id: BranchId,
+    name: ReferenceName,
+    new_change_id: String,
+) -> Result<ChangeReference> {
+    // The reference must be parseable as a remote reference
+    gitbutler_reference::RemoteRefname::from_str(&name)
+        .context("Failed to parse the provided reference")?;
+    let handle = VirtualBranchesHandle::new(ctx.project().gb_dir());
+    // The branch must exist
+    let mut vbranch = handle.get_branch(branch_id)?;
+
+    // Enusre that the commit acutally exists
+    let new_commit =
+        commit_by_branch_id_and_change_id(ctx, &vbranch, &handle, new_change_id.clone())
+            .context(anyhow!("Commit for change_id {} not found", new_change_id))?;
+
+    // Fail early if the commit is not valid
+    validate_commit(&vbranch, new_commit.id(), ctx, &handle)?;
+
+    let reference = vbranch
+        .references
+        .iter_mut()
+        .find(|r| r.name == name)
+        .ok_or(anyhow!(
+            "Reference {} not found for branch {}",
+            name,
+            branch_id
+        ))?;
+    reference.change_id = new_change_id;
+    let new_reference = reference.clone();
+    handle.set_branch(vbranch)?;
+    Ok(new_reference)
+}
+
+/// Pushes a gitbutler branch reference to the remote repository.
+pub fn push_change_reference(
+    ctx: &CommandContext,
+    branch_id: BranchId,
+    name: ReferenceName,
+    with_force: bool,
+    credentials: &Helper,
+) -> Result<()> {
+    let handle = VirtualBranchesHandle::new(ctx.project().gb_dir());
+    let vbranch = handle.get_branch(branch_id)?;
+    let reference = vbranch
+        .references
+        .iter()
+        .find(|r| r.name == name)
+        .ok_or_else(|| anyhow!("Reference {} not found", name))?;
+    let upstream_refname = gitbutler_reference::RemoteRefname::from_str(&reference.name)
```

Comment:
I see, so the `reference.name` is something like `refs/remotes/<remote-without-slash>/<name>`.
Thanks to this constructor, I also think it won't work with multi-slash remotes. Fortunately they are rare, but in the back of my head I want to retire it and replace it with something that gets a list of available remotes as well.

While writing this, I also realize that storing remote names like this will make it difficult to properly deal with renamed remotes. But that's yet another topic.